chronos/asyncsync.nim

```diff
@@ -109,12 +108,15 @@ proc release*(lock: AsyncLock) =
   ## allow exactly one of them to proceed.
   var w: Future[void]
   if lock.locked:
-    lock.locked = false
+    var unlock = true
     while len(lock.waiters) > 0:
       w = lock.waiters.popFirst()
       if not(w.finished()):
+        unlock = false
```

Comment:
is everything that's being called here exception-free, preferably guaranteed by compiler?

Reply:
Conceptually it is exception-free. Formally it's not.